tools/capture_radar.rb

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env ruby
+require File.expand_path("../config/environment", __dir__)
+require 'trollop'
+
+opts = Trollop.options(ARGV) do
+  banner "USAGE:   #{__FILE__} -h <number of hours back to query metrics>\n" \
+         "Example: #{__FILE__} -d <number of days back to query metrics>"
+
+  opt :hours, "Hours", :short => "h", :type => :int, :default => 4
+  opt :days,  "Days",  :short => "d", :type => :int
+  opt :label, "Label", :short => "l", :type => :string, :default => "com.redhat.component"
+end
+Trollop.die :hours, "is required" unless opts[:hours] || opts[:days_given]
+
+Metric
+class Metric
+  def derived_cpu_total_cores_used
+    return if cpu_usage_rate_average.nil? || derived_vm_numvcpus.nil? || derived_vm_numvcpus == 0
+    (cpu_usage_rate_average * derived_vm_numvcpus) / 100.0
+  end
+end
+
+CustomAttribute
+class CustomAttribute
+  has_many   :metric_rollups, :as => :resource
+end
+
+TIME_RANGE = if opts[:days_given]
+               [opts[:days].days.ago.utc.beginning_of_hour..Time.now.utc.end_of_hour]
+             else
+               [opts[:hours].hours.ago.utc.beginning_of_hour..Time.now.utc.end_of_hour]
+             end
+
+labels = CustomAttribute.where(:section => ['labels', 'docker_labels'], :resource_type => ['ContainerImage', 'ContainerGroup'], :name => opts[:label])
+
+# Question - If a container is short lived, will we still be able to find it here?
+labels_containers = labels.each_with_object(Hash.new { |h,k| h[k] = [] }) { |l, h| h[l] += l.resource.containers.to_a }
+
+# Group metrics by label/project
+labels_metrics = Hash.new { |h,k| h[k] = [] }
+labels_containers.each do |l, containers|
+  containers.each do |c|
+    labels_metrics[[l, c.container_project]] += c.metrics.where(:timestamp => TIME_RANGE).to_a # This should be scoped to time range for the hour
+  end
+end
+
+# Calculate the max total (of all containers running with the same label/project) CPU cores for each hour of metrics
+labels_maxes = Hash.new { |h,k| h[k] = {} }
+labels_metrics.each do |label_project, metrics|
+  # Group the metrics for this label/project by timestamp so that we can determine the sum of CPU cores usage for
+  # each 20 interval and use that for comparing the max usages for the hour
+  metrics.group_by { |m| m.timestamp }.each do |ts, metrics|
+    hour_ts = ts.beginning_of_hour
+    cpu_cores = metrics.map { |m| m.derived_cpu_total_cores_used }.sum
+
+    labels_maxes[label_project][hour_ts] ||= { :cpu_usage_rate_average => 0 }
+    labels_maxes[label_project][hour_ts][:cpu_usage_rate_average] = cpu_cores if cpu_cores > labels_maxes[label_project][hour_ts][:cpu_usage_rate_average]
+  end
+end
```

Comment:
@Fryguy @gtanzillo @lpichler so all above can be replaced by 1 query, which computes in 34 seconds and takes around 150MB in Ruby process and 150MB in postgre process

```ruby
def quote(value)
  ActiveRecord::Base.connection.quote(value)
end

def get_hourly_maxes_per_group(label_name)
  sums_query = Metric.where(:resource_type => "Container")
                 .joins("INNER JOIN containers ON metrics.resource_type = 'Container' AND metrics.resource_id = containers.id")
                 .joins("INNER JOIN container_groups ON container_groups.id = containers.container_group_id")
                 .joins("INNER JOIN container_images ON container_images.id = containers.container_image_id")
                 .joins("INNER JOIN custom_attributes ON custom_attributes.name = #{quote(label_name)} AND "\
                    "custom_attributes.resource_type = 'ContainerImage' AND custom_attributes.resource_id = container_images.id")
                 .select("custom_attributes.name as label_name, container_groups.container_project_id as container_project_id, metrics.timestamp as timestamp, "\
                    "sum((metrics.cpu_usage_rate_average * metrics.derived_vm_numvcpus) / 100.0) AS sum_used_cores, "\
                    "count(*) AS containers_in_group")
                 .group("custom_attributes.name, container_groups.container_project_id, metrics.timestamp")
                 .order("custom_attributes.name, container_groups.container_project_id")

  maxes_query = <<-SQL
    WITH sums AS (
      #{sums_query.to_sql}
    )
    SELECT sums.label_name, container_projects.name as container_project_name, date_trunc('hour', sums.timestamp) as hourly_timestamp, max(sums.sum_used_cores) as max_sum_used_cores
      FROM sums
        INNER JOIN container_projects ON container_projects.id = sums.container_project_id
        GROUP BY sums.label_name, container_projects.name, date_trunc('hour', sums.timestamp)
  SQL

  ActiveRecord::Base.connection.execute(maxes_query).to_a
end
```

benefit is that the query goes through only valid metrics, the old way would fetch all archived records too, so it would grow massively both in ram and time.

result looks like
```ruby
result = get_hourly_maxes_per_group("label_name_1")
(byebug) pp result[0..10]
[{"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_1",
  "hourly_timestamp"=>"2018-01-17 11:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_1",
  "hourly_timestamp"=>"2018-01-17 12:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_1",
  "hourly_timestamp"=>"2018-01-17 13:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_1",
  "hourly_timestamp"=>"2018-01-17 14:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_1",
  "hourly_timestamp"=>"2018-01-17 15:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_10",
  "hourly_timestamp"=>"2018-01-17 11:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_10",
  "hourly_timestamp"=>"2018-01-17 12:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_10",
  "hourly_timestamp"=>"2018-01-17 13:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_10",
  "hourly_timestamp"=>"2018-01-17 14:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_10",
  "hourly_timestamp"=>"2018-01-17 15:00:00",
  "max_sum_used_cores"=>560.000000000001},
 {"label_name"=>"label_name_1",
  "container_project_name"=>"container_project_name_100",
  "hourly_timestamp"=>"2018-01-17 11:00:00",
  "max_sum_used_cores"=>560.000000000001}]
```